src/controller/controller.c

```diff
@@ -766,6 +766,200 @@ static int controller_method_list_units(sd_bus_message *m, void *userdata, UNUSE
 }
 #pragma GCC diagnostic pop
 
+/************************************************************************
+ ***** org.eclipse.bluechi.Controller.ListUnitFiles **************
+ ************************************************************************/
+
+typedef struct ListUnitFilesRequest {
+        sd_bus_message *request_message;
+
+        int n_done;
+        int n_sub_req;
+        struct {
+                Node *node;
+                sd_bus_message *m;
+                AgentRequest *agent_req;
+        } sub_req[0];
+} ListUnitFilesRequest;
+
+static void list_unit_files_request_free(ListUnitFilesRequest *req) {
+        sd_bus_message_unref(req->request_message);
+
+        for (int i = 0; i < req->n_sub_req; i++) {
+                node_unrefp(&req->sub_req[i].node);
+                sd_bus_message_unrefp(&req->sub_req[i].m);
+                agent_request_unrefp(&req->sub_req[i].agent_req);
+        }
+
+        free(req);
+}
+
+static void list_unit_files_request_freep(ListUnitFilesRequest **reqp) {
+        if (reqp && *reqp) {
+                list_unit_files_request_free(*reqp);
+                *reqp = NULL;
+        }
+}
+
+#define _cleanup_list_unit_files_request_ _cleanup_(list_unit_files_request_freep)
```

Comment:
`ListUnitFilesRequest` is an exact copy of `ListUnitsRequest`. So I think we can reduce the duplication here and make the code more concise. 
Lets remove the `ListUnitFilesRequest`, rename `ListUnitsRequest` to `AgentRequestList` and generalize the `maybe_done` and `done` part of handling the replies - with passing the `encoding` function as parameter.

Reply:
Added to the GH issue #889 so we can tackle this refactoring later.